llvm/lib/Transforms/Utils/BuildBuiltins.cpp

```diff
@@ -0,0 +1,913 @@
+    PHINode *Result =
+        Builder.CreatePHI(Builder.getInt1Ty(), 3,
+                          Name + "." + getBuiltinSig() + ".failorder.success");
+    IRBuilderBase::InsertPoint ContIP = Builder.saveIP();
+
+    // Emit all the different atomics
+    Builder.SetInsertPoint(MonotonicBB);
+    Value *MonotonicResult =
+        emitInst(IsWeak, SuccessMemorder, AtomicOrdering::Monotonic);
+    Builder.CreateBr(ContBB);
+    Result->addIncoming(MonotonicResult, Builder.GetInsertBlock());
+
+    Builder.SetInsertPoint(AcquireBB);
+    Value *AcquireResult =
+        emitInst(IsWeak, SuccessMemorder, AtomicOrdering::Acquire);
+    Builder.CreateBr(ContBB);
+    Result->addIncoming(AcquireResult, Builder.GetInsertBlock());
+
+    Builder.SetInsertPoint(SeqCstBB);
+    Value *SeqCstResult = emitInst(IsWeak, SuccessMemorder,
+                                   AtomicOrdering::SequentiallyConsistent);
+    Builder.CreateBr(ContBB);
+    Result->addIncoming(SeqCstResult, Builder.GetInsertBlock());
+
+    Builder.restoreIP(ContIP);
+    return Result;
+  };
+
+  Value *emitSuccessMemorderSwitch(bool IsWeak) {
+    if (SuccessMemorderConst)
+      return emitFailureMemorderSwitch(IsWeak, *SuccessMemorderConst);
+
+    Type *BoolTy = Builder.getInt1Ty();
+
+    // Create all the relevant BB's
+    BasicBlock *ContBB =
+        splitBB(Builder, /*CreateBranch=*/false,
+                Name + "." + getBuiltinSig() + ".memorder.continue");
+    BasicBlock *MonotonicBB = createBasicBlock("monotonic");
+    BasicBlock *AcquireBB =
+        supportsAcquireOrdering() ? createBasicBlock("acquire") : nullptr;
+    BasicBlock *ReleaseBB =
+        supportsReleaseOrdering() ? createBasicBlock("release") : nullptr;
+    BasicBlock *AcqRelBB =
+        supportsAcquireOrdering() && supportsReleaseOrdering()
+            ? createBasicBlock("acqrel")
+            : nullptr;
+    BasicBlock *SeqCstBB = createBasicBlock("seqcst");
+
+    // Create the switch for the split
+    // MonotonicBB is arbitrarily chosen as the default case; in practice,
+    // this doesn't matter unless someone is crazy enough to use something
+    // that doesn't fold to a constant for the ordering.
+    IntegerType *IntTy = getIntTy(Builder, EmitOptions.TLI);
+    Value *Order = Builder.CreateIntCast(SuccessMemorderCABI, IntTy, false);
+    SwitchInst *SI = Builder.CreateSwitch(Order, MonotonicBB);
+
+    // TODO: No PHI if operation cannot fail
+    Builder.SetInsertPoint(ContBB, ContBB->begin());
+    PHINode *Result = Builder.CreatePHI(
+        BoolTy, 5, Name + "." + getBuiltinSig() + ".memorder.success");
+    IRBuilderBase::InsertPoint ContIP = Builder.saveIP();
+
+    // Emit all the different atomics
+    Builder.SetInsertPoint(MonotonicBB);
+    Value *MonotonicResult =
+        emitFailureMemorderSwitch(IsWeak, AtomicOrdering::Monotonic);
+    Builder.CreateBr(ContBB);
+    Result->addIncoming(MonotonicResult, Builder.GetInsertBlock());
+
+    if (AcquireBB) {
```

Comment:
We can perhaps wrap the following statements:

```
Builder.SetInsertPoint(AcquireBB);
Value *AcquireResult =
          emitFailureMemorderSwitch(IsWeak, AtomicOrdering::Acquire);
Builder.CreateBr(ContBB);
Result->addIncoming(AcquireResult, Builder.GetInsertBlock());
```

into a lambda with generic parameters (a basic block and a `AtomicOrderingCABI` value) and reuse in these if conditions.